### PR TITLE
Clones object for EntryProcessing if an index exists 

### PR DESCRIPTION
**Pull request by tombujok**
Solves #9798

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapContainer.java [hunk 1]
```diff
@@ -285,6 +286,10 @@ public InterceptorRegistry getInterceptorRegistry() {
     // callback called when the MapContainer is de-registered from MapService and destroyed - basically on map-destroy
     public void onDestroy() {
     }
+
+    public boolean shouldCloneOnEntryProcessing() {
+        return getIndexes().hasIndex() & OBJECT.equals(mapConfig.getInMemoryFormat());
+    }
 }
```
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapContainer.java]: I know that this is an imperfect location for this method but it's the only common context for all the calls that exists right now. I could also create an EntryProcessorUtils class, but seemed an overkill for now. Feel free to chime in.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryOperation.java [hunk 2]
```diff
@@ -61,29 +61,33 @@ public void innerBeforeRun() throws Exception {
...
-            Map.Entry entry = createMapEntry(dataKey, oldValue);
+            Map.Entry entry = createMapEntry(dataKey, value);
             Data response = process(entry);
             if (response != null) {
                 responses.add(dataKey, response);
             }
 
             // first call noOp, other if checks below depends on it.
-            if (noOp(entry, oldValue, now)) {
+            if (noOp(entry, value, now)) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryOperation.java]: Shouldn't this be `oldValue`? Same with `entryRemoved`

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryBackupOperation.java [hunk 3]
```diff
@@ -42,21 +43,24 @@ public PartitionWideEntryBackupOperation(String name, EntryBackupProcessor backu
...
+            if (!applyPredicate(dataKey, value)) {
                 continue;
             }
 
-            Map.Entry entry = createMapEntry(dataKey, oldValue);
+            Map.Entry entry = createMapEntry(dataKey, value);
             processBackup(entry);
 
-            if (noOp(entry, oldValue)) {
+            if (noOp(entry, value)) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryBackupOperation.java]: `oldValue`?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryOperation.java (hunk 2 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryOperation.java]: Thx, fixed.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryBackupOperation.java (hunk 3 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryBackupOperation.java]: Thx, fixed.

**Comment by jerrinot** [comment]
+1 from me.

**Pull request merged**